test/CachedQuery.test.ts

```diff
@@ -1,311 +1,235 @@
-          fetch,
-          1,
-          refetch
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isRight()).toBe(true);
-        const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(2);
-        expect(result2.isRight()).toBe(true);
-        expect(result2.getOrElse('') !== result.getOrElse('')).toBe(true);
+        const fetch = makeTestFetch([
+          (n: number) => taskEither.of(n * 2),
+          (n: number) => taskEither.of(n / 2)
+        ]);
+        const query = queryShallow(fetch, refetch);
+        const result = await query.run(2).run();
+        expect(result).toEqual(right(4));
+        const result2 = await query.run(2).run();
+        expect(result2).toEqual(right(1));
+        fetch.assertExhausted();
       });
 
       it('dopo avere eseguito run con successo una volta, viene chiamata di nuovo run e la fetch esegue nuovamente fallendo e ritornando il fallimento', async () => {
-        const fetch = new StatefulFetch({
-          order: 'successFirst',
-          resultType: 'alwaysDifferent'
-        }).fetch;
-        const { fetchFunction, result, query } = await runQueryShallow(
-          fetch,
-          1,
-          refetch
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isRight()).toBe(true);
+        const fetch = makeTestFetch([
+          (n: number) => taskEither.of(n * 2),
+          () => fromLeft('nope')
+        ]);
+        const query = queryShallow(fetch, refetch);
+        const result = await query.run(1).run();
+        expect(result).toEqual(right(2));
         const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(2);
-        expect(result2.isLeft()).toBe(true);
+        expect(result2).toEqual(left('nope'));
+        fetch.assertExhausted();
       });
 
       it('dopo avere eseguito run con fallimento una volta, viene chiamata di nuovo run e la fetch esegue nuovamente con successo ritornando il nuovo valore', async () => {
-        const fetch = new StatefulFetch({
-          order: 'failureFirst',
-          resultType: 'alwaysDifferent'
-        }).fetch;
-        const { fetchFunction, result, query } = await runQueryShallow(
-          fetch,
-          1,
-          refetch
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isLeft()).toBe(true);
+        const fetch = makeTestFetch([
+          () => fromLeft('nope'),
+          (n: number) => taskEither.of(n * 2)
+        ]);
+        const query = queryShallow(fetch, refetch);
+        const result = await query.run(1).run();
+        expect(result).toEqual(left('nope'));
         const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(2);
-        expect(result2.isRight()).toBe(true);
+        expect(result2).toEqual(right(2));
+        fetch.assertExhausted();
```

Comment:
by testing both of these, aren't we actually testing the util? This was wrong also with the previous implementation I think

Reply:
You mean testing `result` and `result2`? I'd still test `result2` (the final expected result) but probably not `result` as discussed above, since it is part of the prerequisite which is something tested e.g. in the very first test above

Reply:
I actually meant `expect(result2).toEqual(right(2));` and `fetch.assertExhausted();` isn't one implying the other given the util implementation?

Reply:
Yes, I think you are right. Do you see any tests where this could be false? If not, I'd remove the tests on final results too (and keep `assertExhausted` with a better name that being static is easier to copy paste in each test)

Reply:
I think it can be safely removed 👍

Reply:
Actually, rethinking about this: by checking the value of the final result we verify that `run()` per se does not mess in any way with the return value from the test fetch function. The fact that it just returns it internally is an implementation detail. So I still think removing the intermediate results is OK (since we are already testing them in previous tests) but the new results that are checked in each test (only the final one in this case) have a reason to be tested. So I would keep the check